RFC-007-SWAP-Ethereum-Basic.md

```diff
@@ -0,0 +1,291 @@
+| refund_identity | The `address` of the refunding party                                        |
+| expiry          | The absolute UNIX timestamp in seconds after which the HTLC can be refunded |
+
+### Contract
+
+This RFC defines a single disposable contract as the HTLC.
+It uses `selfdestruct` to release the funds to the intended party.
+
+This approach was chosen over a stateful contract written in solidity so the contracts can be precisely defined, verified and updated.
+
+The contract itself works by hard coding the `redeem_identity` and `refund_identity` into the contract.
+When it is called with the correct `secret` as the calldata it will transfer all the Ether to the `redeem_identity` using a `selfdestruct`.
+When it is called without any calldata after the `expiry` it will transfer all the Ether to the `refund_identity` using a `selfdestruct`.
+
+The contract never checks who is calling the contract so it can be called successfully by anyone.
+The funds will only ever be transferred to the `redeem_identity` or the `refund_identity`.
+
+The contract is compiled from the following EVM assembly:
+
+``` asm
+{
+    // Load received secret size
+    calldatasize
+
+    // Check if secret is zero length
+    iszero
+
+    // If secret is zero length, jump to branch that checks if expiry time has been reached
+    check_expiry
+    jumpi
+
+    // Load expected secret size
+    32
+
+    // Load received secret size
+    calldatasize
+
+    // Compare secret size
+    eq
+    iszero
+
+    // If passed secret is wrong size, jump to exit contract
+    exit
+    jumpi
+
+    // Load secret into memory
+    calldatacopy(0, 0, 32)
+
+    // Hash secret with SHA-256 (pre-compiled contract 0x02)
+    call(72, 0x02, 0, 0, 32, 33, 32)
+
+    // Placeholder for correct secret hash
+    <secret_hash>
+
+    // Load hashed secret from memory
+    mload(33)
+
+    // Compare hashed secret with existing one
+    eq
+
+    // Combine `eq` result with `call` result
+    and
+
+    // Jump to redeem if hashes match
+    redeem
+    jumpi
+
+    // Exit if hashes don't match
+    return(0, 0)
+
+check_expiry:
+    // Timestamp of the current block in seconds since the epoch
+    timestamp
+
+    // Placeholder for refund timestamp
+    <expiry>
+
+    // Compare refund timestamp with current timestamp
+    lt
+
+    // Jump to refund if time is expired
+    refund
+    jumpi
+
+exit:
+    // Exit
+    return(0, 0)
+
+redeem:
+    log1(0, 32, 0xB8CAC300E37F03AD332E581DEA21B2F0B84EAAADC184A295FEF71E81F44A7413) // log keccak256("Redeemed()")
+    selfdestruct(<redeem_identity>)
+
+refund:
+    log1(0, 0, 0x5D26862916391BF49478B2F5103B0720A842B45EF145A268F2CD1FB2AED55178) // log keccak256("Refunded()")
+    selfdestruct(<refund_identity>)
+}
+
+```
+
+The following is the contract deployment code encoded template as hex which will deploy the above contract when the placeholder values (set to `00`s) are replaced with the parameters.
```

Comment:
Maybe we should leave the `<placeholder>` in there like that since it is totally valid for `00` to appear in the contract?

Reply:
I like the idea. The placeholders have to be the right length obviously.
The thing is I wouldn't expect them to store the placeholders as hex but as bytes and overwrite the bytes (and there's no such thing as an invalid byte sequence!)